### PR TITLE
WiP: sparse

### PR DESCRIPTION
**Pull request by dim**

### CODE REVIEW

File: hllplus/hllplus.go [hunk 1]
```diff
@@ -82,14 +92,23 @@ func (s *HLL) Merge(other *HLL) {
 		return
 	}
 
+	// FIXME: allow sparse merge
```
Comment (dim) [inline hllplus/hllplus.go:95]: No need to fix this as part of this PR, we can do the sparse/sparse + sparse/dense merging separately later

File: hllplus/hllplus.go [hunk 2]
```diff
@@ -191,17 +216,21 @@ func (s *HLL) Downgrade(precision, sparsePrecision uint8) error {
 	return nil
 }
 
+func (s *HLL) normalize() {
```
Comment (dim) [inline hllplus/hllplus.go:219]: This is the method that is not ready yet, at the moment it just switches from sparse to dense but it does not migrate existing/stored values, it's unfortunately a little tricky, but I will point you at the relevant code.

File: hllplus/hllplus_test.go [hunk 3]
```diff
@@ -93,9 +128,41 @@ var _ = Describe("HLL", func() {
...
+			}
+			Expect(subject.IsSparse()).To(BeTrue())
+			Expect(subject.Estimate()).To(Equal(int64(exp)))
+		},
+		Entry("p=23", 23, 150823),
+		Entry("p=24", 24, 150144),
+		Entry("p=25", 25, 150012),
+	)
+
+	FIt("should normalize", func() {
```
Comment (dim) [inline hllplus/hllplus_test.go:150]: This is the test that is currently failing/needs fixing, pls ignore the failing tests in the root package for now

File: hllplus/sparse.go [hunk 4]
```diff
@@ -1 +1,256 @@
...
+var deltaSlicePool sync.Pool
+
+// Delta encoded slice of uint32s.
+type deltaSlice struct {
+	nums uvarintSlice
+	last uint32
+	size int
+}
+
+func recycleDeltaSlice(size int) *deltaSlice {
```
Comment (mxmCherry) [inline hllplus/sparse.go:203]: Nit: I find `acquire...` more common name for pooled data, why `recycle...`?

File: hllplus/sparse.go [hunk 5]
```diff
@@ -1 +1,256 @@
...
+	s.size++
+}
+
+func (s *deltaSlice) Iterate(fn func(uint32)) {
+	var last uint32
+	s.nums.Iterate(func(u uint32) {
+		x := u + last
+		fn(x)
+		last = x
+	})
```
Comment (mxmCherry) [inline hllplus/sparse.go:255]: @dim just confirming that this is right ^^

(any concerns about overflows?)

File: hllplus/sparse.go [hunk 6]
```diff
@@ -1 +1,256 @@
...
+	}
+
+	t := make(uint32Set, len(s))
+	for n := range s {
+		t.Add(n)
+	}
+	return t
+}
+
+func (s uint32Set) Flush() []uint32 {
```
[the remark this answers is not in the document]
Comment (dim) [reply hllplus/sparse.go:151]: I called it Slice originally, the reality is that we buffer and eventually `Flush`, i.e. the buffer is emptied when this method is called

File: hllplus/sparse.go (hunk 4 above)
Comment (dim) [reply hllplus/sparse.go:203]: naming taken from the Java lib, I like it :shrug:

File: hllplus/sparse.go (hunk 5 above)
Comment (dim) [reply hllplus/sparse.go:255]: No, not concern, the maximum value that can be stored in this "slice" is `(1 << p)` and our `p` can never be > 25 (so it definitely fits a uint32). The delta encoded slice is always sorted first so the values are always ascending

**Comment by mxmCherry** [comment]
@dim merge?

**Pull request merged**